File: Day230/splitLinkedListInPairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        ListNode arr[] = new ListNode[k];
        ListNode temp = head;
        int count = 0;
        while (temp != null) {
            count++;
            temp = temp.next;
        }
        temp = head;
        int rem = count % k;
        int c = count / k;
        int point  = 0;
        
        while (temp != null) {
            arr[point++] = temp;
            int len = c - 1 + ((rem-- > 0) ? 1: 0);
            for (int i=0; i<len; i++) {
                temp = temp.next;
            }
            ListNode t = temp.next;
            temp.next = null;
            temp = t;
        }
        return arr;
    }
}

File: Day230/splitLinkedListInPairs.go

//   Definition for singly-linked list.
type ListNode struct {
     Val int
      Next *ListNode
  }
 
 func splitListToParts(head *ListNode, k int) []*ListNode {
    arr := make([]*ListNode, k)
    temp := head
    count := 0
    for temp != nil {
        count++
        temp = temp.Next
    }
    temp = head
    rem := count % k
    c := count / k
    point := 0
    for temp != nil {
        arr[point] = temp
        point++
        l := c - 1
        if rem > 0 {
            l += 1
        }
        rem--
        for i:=0; i<l; i++ {
            temp = temp.Next
        }
        t := temp.Next
        temp.Next = nil
        temp = t
    }
    return arr
}